src/serializer/ResidualHeapSerializer.js

```diff
@@ -866,16 +869,22 @@ export class ResidualHeapSerializer {
       case "Number":
         let numberData = val.$NumberData;
         invariant(numberData !== undefined);
+        numberData.throwIfNotConcreteNumber();
```

Comment:
I think that change is out of scope of this issue. The goal of this pull request is to get toString() working on an abstract value. I will file another issue for your suggestion.
Btw: even we implemented in serializer, it won't work E2E; because the following code will fail in ToNumber() call in to.js.
```
let x = global.__abstract ? __abstract("number", 42) : 42;
let y = Number(x);
```